epochStart/metachain/economics.go

```diff
@@ -329,12 +346,91 @@ func (e *economics) adjustRewardsPerBlockWithLeaderPercentage(
 }
 
 // compute inflation rate from genesisTotalSupply and economics settings for that year
-func (e *economics) computeInflationRate(currentRound uint64, currentEpoch uint32) float64 {
+func (e *economics) computeInflationRateLegacy(currentRound uint64, currentEpoch uint32) (float64, error) {
 	roundsPerDay := common.ComputeRoundsPerDay(e.roundTime.TimeDuration(), e.enableEpochsHandler, currentEpoch)
+
 	roundsPerYear := numberOfDaysInYear * roundsPerDay
 	yearsIndex := uint32(currentRound/roundsPerYear) + 1
 
-	return e.rewardsHandler.MaxInflationRate(yearsIndex)
+	return e.rewardsHandler.MaxInflationRate(yearsIndex), nil
+}
+
+func (e *economics) computeInflationRate(currentRound uint64, currentEpoch uint32) (float64, error) {
+	supernovaInEpochActivated := e.enableEpochsHandler.IsFlagEnabledInEpoch(common.SupernovaFlag, currentEpoch)
+	if !supernovaInEpochActivated {
+		return e.computeInflationRateLegacy(currentRound, currentEpoch)
+	}
+
+	currentYear := currentEpoch/uint32(numberOfDaysInYear) + 1
+	supernovaActivationYear, err := e.getSupernovaActivationYear(currentEpoch)
+	if err != nil {
+		return 0, err
+	}
+
+	if currentYear == supernovaActivationYear {
+		return e.computeInflationRateInSupernovaYear(currentYear)
+	}
+
+	return e.computeInflationRateAfterSupernovaYear(currentRound, currentEpoch, supernovaActivationYear)
+}
+
+func (e *economics) getPrevSupernovaActivationEpoch(supernovaActivationEpoch uint32) uint32 {
+	if supernovaActivationEpoch == 0 {
+		return supernovaActivationEpoch
+	}
+
+	return supernovaActivationEpoch - 1
+}
+
+func (e *economics) getSupernovaActivationYear(currentEpoch uint32) (uint32, error) {
+	supernovaActivationRound := e.enableRoundsHandler.GetActivationRound(common.SupernovaRoundFlag)
+	supernovaActivationEpoch := e.enableEpochsHandler.GetActivationEpoch(common.SupernovaFlag)
+
+	supernovaPrevActivationEpoch := e.getPrevSupernovaActivationEpoch(supernovaActivationEpoch)
+	chainParametersUntilSupernova, err := e.chainParamsHandler.ChainParametersForEpoch(supernovaPrevActivationEpoch)
+	if err != nil {
+		return 0, err
+	}
+	roundDuration := chainParametersUntilSupernova.RoundDuration
+	if roundDuration <= 0 {
+		return 0, errors.ErrInvalidRoundDuration
+	}
+
+	roundsPerDay := common.ComputeRoundsPerDay(time.Duration(roundDuration)*time.Millisecond, e.enableEpochsHandler, supernovaPrevActivationEpoch)
```

Comment:
general idea in the start of epoch economics computation is that you are computing the rewards and inflation for the previous day/epoch.

Here it seems that the current epoch you are using in the computation is the new epoch (taken from the epoch start meta), so you might end up computing a different number of rounds per epoch.

Although it seems this is only locally used, and doesn't affect too much the end result (as the inflation per day during the year is the most of the time the same, no matter if you are computing for the previous day or the current day), but you might be entering in the next year with the inflation one day too soon.

Can you make sure the computation aligns with the idea of calculating the rewards for the previous epoch?

Reply:
done